README.md
# install_plugin

[![Build Status](https://travis-ci.org/hui-z/flutter_install_plugin.svg?branch=master)](https://travis-ci.org/hui-z/flutter_install_plugin#)
[![pub package](https://img.shields.io/pub/v/install_plugin.svg)](https://pub.dartlang.org/packages/install_plugin)
[![license](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/hui-z/flutter_install_plugin/blob/master/LICENSE)

We use the `install_plugin` plugin to install apk for android; and using url to go to app store for iOS.

## Usage

To use this plugin, add `install_plugin` as a dependency in your pubspec.yaml file. For example:
```yaml
dependencies:
  install_plugin: '^2.1.0'
```

## iOS
Your project need create with swift.

##  Android
You need to request permission for READ_EXTERNAL_STORAGE to read the apk file. You can handle the storage permission using [flutter_permission_handler](https://github.com/BaseflowIT/flutter-permission-handler).
```
 <!-- read permissions for external storage -->
 <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
```
In Android version >= 8.0 , You need to request permission for REQUEST_INSTALL_PACKAGES to install the apk file
 ```
 <!-- installation package permissions -->
 <uses-permission android:name="android.permission.REQUEST_INSTALL_PACKAGES" />
 ```
In Android version <= 6.0 , You need to request permission for WRITE_EXTERNAL_STORAGE to copy the apk from the app private location to the download directory
 ```
 <!-- write permissions for external storage -->
 <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
 ```

## Example
install apk from the internet 
``` dart
  _networkInstallApk() async {
    if (_progressValue != 0 && _progressValue < 1) {
      _showResMsg("Wait a moment, downloading");
      return;
    }

    _progressValue = 0.0;
    var appDocDir = await getTemporaryDirectory();
    String savePath = appDocDir.path + "/takeaway_phone_release_1.apk";
    String fileUrl =
        "https://s3.cn-north-1.amazonaws.com.cn/mtab.kezaihui.com/apk/takeaway_phone_release_1.apk";
    await Dio().download(fileUrl, savePath, onReceiveProgress: (count, total) {
      final value = count / total;
      if (_progressValue != value) {
        setState(() {
          if (_progressValue < 1.0) {
            _progressValue = count / total;
          } else {
            _progressValue = 0.0;
          }
        });
        print((_progressValue * 100).toStringAsFixed(0) + "%");
      }
    });

    final res = await InstallPlugin.install(savePath);
    _showResMsg(
        "install apk ${res['isSuccess'] == true ? 'success' : 'fail:${res['errorMessage'] ?? ''}'}");
  }
```

install apk from the local storage
``` dart
  _localInstallApk() async {
    FilePickerResult? result = await FilePicker.platform.pickFiles();
    if (result != null) {
      final res = await InstallPlugin.install(result.files.single.path ?? '');
      _showResMsg(
          "install apk ${res['isSuccess'] == true ? 'success' : 'fail:${res['errorMessage'] ?? ''}'}");
    } else {
      // User canceled the picker
      _showResMsg("User canceled the picker apk");
    }
  }
```

Go to AppStore , example appStore url : https://itunes.apple.com/cn/app/%E5%86%8D%E6%83%A0%E5%90%88%E4%BC%99%E4%BA%BA/id1375433239?l=zh&ls=1&mt=8
``` dart
  _gotoAppStore(String url) async {
    url = url.isEmpty ? _defaultUrl : url;
    final res = await InstallPlugin.install(url);
    _showResMsg(
        "go to appstroe ${res['isSuccess'] == true ? 'success' : 'fail:${res['errorMessage'] ?? ''}'}");
  }
```


CHANGELOG.md
## 2.1.0

* Support low version Android installation (tested 5, 6, 7, 8, 9, 10, 11, 12, 13, the version lower than 5 has not been tested, if there is any problem, please report to me)
* Added installation support after downloading apk from the network
* Optimize Android installation code
* Optimize Apple's jump code

## 2.0.1

* fix issue 15,17

## 2.0.0

* Breaking change. Migrate from the deprecated original Android Support Library to AndroidX.This shouldn't result in any functional changes, but it requires any Android apps using this plugin to also migrate if they're using the original support library.

## 1.2.0

* Upgrade kotlin(1.2.71) and gradle build plugin version(3.2.1).

## 1.1.0

* Remove `getPlatformVersion` function.

## 1.0.0

* Add CHANGELOG and README.
* Remove unused code.
* Update plugin homepage url.

## 0.0.2

* Fix the bug that Pixel Device install apk fail.
* Optimize the user experience when app asking for `Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES` permission for Android device above ANDROID.O.

## 0.0.1

* Add `installApk` function for android to install apk, supporting ANDROID.O.
* Add `gotoAppStore` function for iOS to launch a url to App Store.

example/lib/main.dart
import 'dart:io';

import 'package:dio/dio.dart';
import 'package:file_picker/file_picker.dart';
import 'package:flutter/material.dart';
import 'package:install_plugin/install_plugin.dart';
import 'package:path_provider/path_provider.dart';

import 'utils.dart';

void main() => runApp(new MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => new _MyAppState();
}

class _MyAppState extends State<MyApp> {
  /// 默认使用了再惠合伙人的下载地址
  static const _defaultUrl =
      'https://itunes.apple.com/cn/app/%E5%86%8D%E6%83%A0%E5%90%88%E4%BC%99%E4%BA%BA/id1375433239?l=zh&ls=1&mt=8';
  TextEditingController _textEditingController = TextEditingController();
  String _appUrl = '';
  double _progressValue = 0.0;

  @override
  void initState() {
    super.initState();
    _textEditingController.text = _defaultUrl;
    PermissionUtil.requestAll();
  }

  @override
  void dispose() {
    _textEditingController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Container(
          alignment: Alignment.center,
          child: Platform.isAndroid
              ? Column(
                  crossAxisAlignment: CrossAxisAlignment.center,
                  children: [
                    Padding(
                      padding:
                          const EdgeInsets.only(top: 30, left: 16.0, right: 16),
                      child: LinearProgressIndicator(
                        value: _progressValue,
                        backgroundColor: Colors.grey,
                        valueColor: AlwaysStoppedAnimation<Color>(Colors.blue),
                      ),
                    ),
                    Padding(
                      padding: const EdgeInsets.only(
                          top: 16, left: 16.0, right: 16, bottom: 16),
                      child: Row(
                        mainAxisAlignment: MainAxisAlignment.end,
                        children: [
                          Text(
                              'downloading ${(_progressValue * 100).toStringAsFixed(0)} %')
                        ],
                      ),
                    ),
                    ElevatedButton(
                        onPressed: () => _networkInstallApk(),
                        child: Text('network install apk')),
                    SizedBox(height: 10),
                    ElevatedButton(
                        onPressed: () => _localInstallApk(),
                        child: Text('local install apk')),
                  ],
                )
              : Column(
                  crossAxisAlignment: CrossAxisAlignment.center,
                  children: <Widget>[
                    Platform.isIOS
                        ? TextField(
                            controller: _textEditingController,
                            decoration: InputDecoration(
                                hintText: 'URL for app store to launch'),
                            onChanged: (url) => _appUrl = url,
                          )
                        : SizedBox(),
                    Platform.isIOS
                        ? ElevatedButton(
                            onPressed: () => _gotoAppStore(_appUrl),
                            child: Text('gotoAppStore'))
                        : SizedBox()
                  ],
                ),
        ),
      ),
    );
  }

  _networkInstallApk() async {
    if (_progressValue != 0 && _progressValue < 1) {
      _showResMsg("Wait a moment, downloading");
      return;
    }

    _progressValue = 0.0;
    var appDocDir = await getTemporaryDirectory();
    String savePath = appDocDir.path + "/zaihui_kylinim_42.apk";
    String fileUrl =
        "https://s3.cn-north-1.amazonaws.com.cn/mtab.kezaihui.com/apk/kylinim/zaihui_kylinim_42.apk";
    await Dio().download(fileUrl, savePath, onReceiveProgress: (count, total) {
      final value = count / total;
      if (_progressValue != value) {
        setState(() {
          if (_progressValue < 1.0) {
            _progressValue = count / total;
          } else {
            _progressValue = 0.0;
          }
        });
        print((_progressValue * 100).toStringAsFixed(0) + "%");
      }
    });

    final res = await InstallPlugin.install(savePath);
    _showResMsg(
        "install apk ${res['isSuccess'] == true ? 'success' : 'fail:${res['errorMessage'] ?? ''}'}");
  }

  _localInstallApk() async {
    FilePickerResult? result = await FilePicker.platform.pickFiles();
    if (result != null) {
      final res = await InstallPlugin.install(result.files.single.path ?? '');
      _showResMsg(
          "install apk ${res['isSuccess'] == true ? 'success' : 'fail:${res['errorMessage'] ?? ''}'}");
    } else {
      // User canceled the picker
      _showResMsg("User canceled the picker apk");
    }
  }

  _gotoAppStore(String url) async {
    url = url.isEmpty ? _defaultUrl : url;
    final res = await InstallPlugin.install(url);
    _showResMsg(
        "go to appstroe ${res['isSuccess'] == true ? 'success' : 'fail:${res['errorMessage'] ?? ''}'}");
  }

  _showResMsg(String msg) {
    print(msg);
    Utils.toast(msg);
  }
}
